sdk/core/core-rest-pipeline/src/policies/retryPolicy.ts

```diff
@@ -0,0 +1,130 @@
+import { RestError } from "../restError";
+import { AbortError } from "@azure/abort-controller";
+
+const retryPolicyLogger = createClientLogger("core-rest-pipeline retryPolicy");
+const DEFAULT_MAX_RETRIES = 3;
+
+/**
+ * The programmatic identifier of the retryPolicy.
+ */
+const retryPolicyName = "retryPolicy";
+
+/**
+ * retryPolicy is a generic policy to enable retrying requests when certain conditions are met
+ */
+export function retryPolicy(...strategies: RetryStrategy[]): PipelinePolicy {
+  return {
+    name: retryPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      let response: PipelineResponse | undefined;
+      let responseError: RestError | undefined;
+      let retryCount = -1;
+
+      const strategyState: RetryStrategyState[] = Array(strategies.length);
+
+      // eslint-disable-next-line no-constant-condition
+      retryRequest: while (true) {
+        retryCount += 1;
+        response = undefined;
+        responseError = undefined;
+
+        try {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Attempting to send request`,
+            request.requestId
+          );
+          response = await next(request);
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received a response from request`,
+            request.requestId
+          );
+        } catch (e) {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received an error from request`,
+            request.requestId
+          );
+          responseError = e as RestError;
+          if (responseError.name !== "RestError") {
+            throw responseError;
+          }
+          response = responseError.response;
+        }
+
+        retryPolicyLogger.info(
+          `Retry ${retryCount}: Processing ${strategies.length} retry strategies.`
+        );
+
+        for (const [i, strategy] of strategies.entries()) {
+          const strategyLogger = strategy.logger || retryPolicyLogger;
+          strategyLogger.info(`Retry ${retryCount}: Processing retry strategy ${strategy.name}.`);
+          let state: RetryStrategyState = {
+            ...strategyState[i],
+            retryCount,
+            response,
+            responseError
+          };
+
+          if (state.retryCount >= (state.maxRetries || DEFAULT_MAX_RETRIES)) {
+            strategyLogger.info(
+              `Maximum retries reached. Returning the last received response, or throwing the last received error.`
+            );
+            if (response) {
+              return response;
+            }
+            throw responseError;
+          }
+
+          if (!strategy.meetsConditions || strategy.meetsConditions(state)) {
+            state = strategy.updateRetryState(state);
+          } else {
+            strategyLogger.error(`Retry ${retryCount}: Does not meet conditions.`);
+            delete state.retryAfterInMs;
+          }
```

Comment:
This feels a bit strange to me, since `updateRetryState` is mandatory, why not have that method return `state` but with `retryAfterInMs` removed, if that is necessary, rather than adding a secondary method that returns false?

Reply:
The idea of meets conditions is to allow retry strategies to fall through depending on the error/response received.

It’s not the same since you might have two strategies that might collide with each other if they can’t fall through depending on a condition.

Reply:
An inverted model would be to separate retry information from the commands returned from the `updateRetryState`.

For example, if the types were:

```ts
export interface RetryInformation {
  response?: PipelineResponse;
  responseError?: RestError;
  retryCount: number;
}

export interface RetryModifiers {
  skipRetry?: boolean;
  maxRetries?: number;
  redirectTo?: string;
  retryAfterInMs?: number;
  throwError?: RestError;
 }

export interface RetryStrategy {
  logger?: AzureLogger;
  name: string;
  retry(retryInformation: RetryInformation): RetryModifiers
 }
 ```

Reply:
How does this look? https://github.com/Azure/azure-sdk-for-js/pull/19078/commits/706de41ce49045e4afaa8559dafcd8204a1e122c